workflow/rules/quant.smk

```diff
@@ -1,28 +1,336 @@
+    output:
+        canonical_mapped_3prime_pos=temp(
+            "results/mapped_3prime_bam/{sample}-{unit}.canonical.mapped.3prime_pos.txt"
+        ),
+    log:
+        "results/logs/mapped_3prime_bam/{sample}-{unit}.mapped.pos.log",
+    conda:
+        "../envs/QC.yaml"
+    script:
+        "../scripts/get-3prime-max-positions.py"
+
+
+rule get_closest_3prime_aligned_pos_bam:
+    input:
+        canonical_mapped_bam="results/canonical_mapped_bam/{sample}-{unit}.canonical.mapped.bam",
+        canonical_mapped_3prime_pos="results/mapped_3prime_bam/{sample}-{unit}.canonical.mapped.3prime_pos.txt",
+    output:
+        canonical_mapped_3prime_bam=temp(
+            "results/canonical_3prime_mapped_bam/{sample}-{unit}.canonical.3prime_mapped.bam"
+        ),
+    log:
+        "results/logs/canonical_3prime_mapped_bam/{sample}-{unit}.canonical.3prime_mapped.log",
+    conda:
+        "../envs/samtools.yaml"
+    shell:
+        "samtools view -R {input.canonical_mapped_3prime_pos} {input.canonical_mapped_bam} -o {output.canonical_mapped_3prime_bam}  2> {log}"
+
+
+rule get_canonical_fastq:
+    input:
+        canonical_3prime_mapped_bam="results/canonical_3prime_mapped_bam/{sample}-{unit}.canonical.3prime_mapped.bam",
+    output:
+        canonical_fastq="results/canonical_reads/{sample}-{unit}.fastq",
+    log:
+        "results/logs/canonical_3prime_mapped_bam/{sample}-{unit}.canonical_3prime_mapped.fastq.log",
+    conda:
+        "../envs/samtools.yaml"
+    shell:
+        "samtools bam2fq {input.canonical_3prime_mapped_bam} > {output.canonical_fastq}  2> {log}"
+
+
+rule get_aligned_pos:
+    input:
+        bam_file="results/kallisto_cdna/{sample}-{unit}",
+    output:
+        aligned_files=temp("results/QC/{sample}-{unit}.aligned.txt"),
+    log:
+        "results/logs/QC/{sample}-{unit}.aligned.log",
+    conda:
+        "../envs/samtools.yaml"
+    shell:
+        "samtools view {input.bam_file}/pseudoalignments.bam | cut -f1,3,4,10,11  > {output} 2> {log}"
+
+
+rule kallisto_samtools_sort:
+    input:
+        "results/kallisto_cdna/{sample}-{unit}",
+    output:
+        temp("results/kallisto-bam-sorted/{sample}-{unit}-pseudoalignments.sorted.bam"),
+    log:
+        "results/logs/QC/{sample}-{unit}.sorted.log",
+    conda:
+        "../envs/samtools.yaml"
+    shell:
+        "samtools sort {input}/pseudoalignments.bam > {output} 2> {log}"
+
+
+rule kallisto_samtools_index:
+    input:
+        "results/kallisto-bam-sorted/{sample}-{unit}-pseudoalignments.sorted.bam",
+    output:
+        temp(
+            "results/kallisto-bam-sorted/{sample}-{unit}-pseudoalignments.sorted.bam.bai"
+        ),
+    log:
+        "results/logs/QC/{sample}-{unit}.sorted.index.log",
+    params:
+        extra="",  # optional params string
+    threads: 4  # This value - 1 will be sent to -@
+    wrapper:
+        "v1.18.3/bio/samtools/index"
+
+
+if is_3prime_experiment and config["experiment"]["3-prime-rna-seq"]["plot-qc"] != "all":
```

Comment:
Please move these into a separate file called `qc_3prime.smk`. Further, the rules are very similar. You can keep the transcript specific version you have below, and for the "all" version, you can simplify the definition via a use rule statement:

```python
use rule get_selected_transcripts_aligned_read_bins as get_aligned_read_bins with:
    output:
        # and define diverging output here
```

Reply:
Noted. Changed.